src/core/block.rs

```diff
@@ -0,0 +1,512 @@
+    btck_TransactionSpentOutputs, btck_block_copy, btck_block_count_transactions,
+    btck_block_create, btck_block_destroy, btck_block_get_hash, btck_block_get_transaction_at,
+    btck_block_hash_destroy, btck_block_spent_outputs_copy, btck_block_spent_outputs_count,
+    btck_block_spent_outputs_destroy, btck_block_spent_outputs_get_transaction_spent_outputs_at,
+    btck_block_to_bytes, btck_block_tree_entry_get_block_hash, btck_block_tree_entry_get_height,
+    btck_block_tree_entry_get_previous, btck_coin_confirmation_height, btck_coin_copy,
+    btck_coin_destroy, btck_coin_get_output, btck_coin_is_coinbase,
+    btck_transaction_spent_outputs_copy, btck_transaction_spent_outputs_count,
+    btck_transaction_spent_outputs_destroy, btck_transaction_spent_outputs_get_coin_at,
+};
+
+use crate::{c_helpers, c_serialize, ChainstateManager, KernelError, TransactionRef, TxOutRef};
+
+/// A type for a Block hash.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct BlockHash {
+    pub hash: [u8; 32],
+}
+
+/// A block tree entry that is tied to a specific [`ChainstateManager`].
+///
+/// Internally the [`ChainstateManager`] keeps an in-memory of the current block
+/// tree once it is loaded. The [`BlockTreeEntry`] points to an entry in this tree.
+/// It is only valid as long as the [`ChainstateManager`] it was retrieved from
+/// remains in scope.
+#[derive(Debug)]
+pub struct BlockTreeEntry<'a> {
+    inner: *const btck_BlockTreeEntry,
+    marker: PhantomData<&'a ChainstateManager>,
+}
+
+unsafe impl Send for BlockTreeEntry<'_> {}
+unsafe impl Sync for BlockTreeEntry<'_> {}
+
+impl<'a> BlockTreeEntry<'a> {
+    pub unsafe fn from_ptr(ptr: *const btck_BlockTreeEntry) -> Self {
+        BlockTreeEntry {
+            inner: ptr,
+            marker: PhantomData,
+        }
+    }
+
+    /// Move to the previous entry in the block tree. E.g. from height n to
+    /// height n-1.
+    pub fn prev(self) -> Option<BlockTreeEntry<'a>> {
+        let inner = unsafe { btck_block_tree_entry_get_previous(self.inner) };
+
+        if inner.is_null() {
+            return None;
+        }
+
+        Some(unsafe { BlockTreeEntry::from_ptr(inner) })
+    }
+
+    /// Returns the current height associated with this BlockTreeEntry.
+    pub fn height(&self) -> i32 {
+        unsafe { btck_block_tree_entry_get_height(self.inner) }
+    }
+
+    /// Returns the current block hash associated with this BlockTreeEntry.
+    pub fn block_hash(&self) -> BlockHash {
+        let hash = unsafe { btck_block_tree_entry_get_block_hash(self.inner) };
+        let res = BlockHash {
+            hash: unsafe { (&*hash).hash },
+        };
+        unsafe { btck_block_hash_destroy(hash) };
+        res
+    }
+
+    pub fn as_ptr(&self) -> *const btck_BlockTreeEntry {
+        self.inner
+    }
+}
+
+impl<'a> Clone for BlockTreeEntry<'a> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a> Copy for BlockTreeEntry<'a> {}
```

Comment:
In commit 86a0f2097991cecb59d83e08da61fbee202dfa3c: I would move this into a separate module, since we use it for things that are not directly block related